File: src/modeling_software-systems_labs/lab2.js
/**
 * Лабораторна робота №2
 * Побудова програми обробки константи типу INTEGER згідно автоматної моделі.
 */
function parseIntegerAutomaton(buff) {
  console.log(`\n======================================`);
  console.log(`Вхідний буфер: "${buff}"`);
  console.log(`======================================`);

  // --- Початковий стан b ---
  let sum = 0; // нульова початкова сума
  let i = 0; // позиція в буфері (в JS індексація з 0)
  let fsign = 0; // прапорець знаку
  let isError = false;

  const max = buff.length;
  const smax = 2147483647; // максимальне значення для 32-бітного цілого числа

  if (max === 0) {
    console.log('-> Стан 4: error (Порожній рядок)');
    return null;
  }

  // --- Стан 1 та 2: Розпізнавання знаку ---
  // Предикат {buff(i)='+'}
  if (buff[i] === '+') {
    console.log("-> Стан 1: {buff(i)='+'} -> перехід до m1");
    i++;
  }
  // Предикат {buff(i)='-'}
  else if (buff[i] === '-') {
    console.log("-> Стан 2: {buff(i)='-'} -> fsign=1, перехід до m1");
    fsign = 1;
    i++;
  }

  // Перевірка, чи залишились символи після знаку
  if (i >= max) {
    console.log('-> Стан 4: error (Відсутні цифри після знаку)');
    return null;
  }

  // --- Цикл розпізнавання (мітка m1) ---
  while (i < max) {
    const char = buff[i];

    // Предикат {buff(i)=cyf} - перевірка на належність до цифр
    if (char >= '0' && char <= '9') {
      // Функція cstod(buff(i)) - перетворення символу в число через ASCII-коди
      const dig = char.charCodeAt(0) - '0'.charCodeAt(0);

      // Формування числа зі старших розрядів
      sum = sum * 10 + dig;
      console.log(`-> Стан 3: Обробка цифри '${char}', поточна sum = ${sum}`);

      // Вкладений предикат {sum > smax}
      if (sum > smax) {
        console.log(`-> Стан 4: error (Перевищення smax = ${smax})`);
        isError = true;
        break; // Вихід згідно моделі обробки помилок
      }
    } else {
      // Предикат {otherwise} у стані розпізнавання цифр
      console.log(`-> Стан 4: error (Неприпустимий символ '${char}')`);
      isError = true;
      break;
    }

    i++; // Перехід на наступну ітерацію (i++)
  }

  // --- Кінцевий стан e ---
  if (!isError) {
    // Предикат {fsign=1}[sum= - sum, stop]
    if (fsign === 1) {
      sum = -sum;
    }
    console.log(`-> Стан e (stop): Успішно! Кінцевий результат = ${sum}`);
    return sum;
  } else {
    console.log('-> Стан e (stop): Роботу припинено через помилку (error).');
    return null;
  }
}

// ==========================================
// Тестування (Перевірка на різних даних)
// ==========================================
parseIntegerAutomaton('2641'); // 1. Нормальне додатне число
parseIntegerAutomaton('-4096'); // 2. Від'ємне число
parseIntegerAutomaton('+128'); // 3. Число з явним плюсом
parseIntegerAutomaton('26A1'); // 4. Помилка: недопустимий символ 'A'
parseIntegerAutomaton('3000000000'); // 5. Помилка: перевищення smax

File: src/utils/c++sandbox/lab2.cpp
#include <iostream>
#include <string>
#include <climits> // Для перевірки smax (INT_MAX)

using namespace std;

void processInteger() {
    string buff;
    cout << "Введіть число (INTEGER) у символьному вигляді: ";
    cin >> buff;

    // Стан П: Ініціалізація
    long long sum = 0; // Використовуємо long long для перевірки переповнення
    int i = 0;
    int fsign = 0; // 0 - позитивне, 1 - негативне
    int max_len = buff.length();

    // Перевірка на порожній рядок
    if (max_len == 0) {
        cout << "Помилка: Порожній ввід" << endl;
        return;
    }

    // Стан 1 та 2: Обробка знаку
    if (buff[i] == '+') {
        i++; // Просто переходимо до наступного символу
    } else if (buff[i] == '-') {
        fsign = 1;
        i++;
    }

    // Перевірка, чи є цифри після знаку
    if (i >= max_len) {
        cout << "Стан 4 (Error): Очікувалася цифра після знаку." << endl;
        return;
    }

    // Стан 3: Обробка цифр
    while (i < max_len) {
        char current = buff[i];

        // Перевірка, чи є символ цифрою (cyf)
        if (current >= '0' && current <= '9') {
            int dig = current - '0'; // cstod - перетворення символу в цифру
            
            sum = sum * 10 + dig;

            // Перевірка на переповнення {sum > smax}
            if ((fsign == 0 && sum > INT_MAX) || (fsign == 1 && (-sum) < INT_MIN)) {
                cout << "Стан 4 (Error): Перевищення максимального значення (smax)!" << endl;
                return;
            }
        } else {
            // Стан 4: Недопустимий символ
            cout << "Стан 4 (Error): Неприпустимий символ '" << current << "'." << endl;
            return;
        }
        i++;
    }

    // Стан Е: Кінцева обробка
    int finalResult = (int)sum;
    if (fsign == 1) {
        finalResult = -finalResult;
    }

    cout << "Результат у внутрішньому форматі: " << finalResult << endl;
}

int main() {
    processInteger();
    return 0;
}